src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -0,0 +1,302 @@
+package org.broadinstitute.dsde.rawls.expressions
+
+import _root_.slick.dbio
+import org.broadinstitute.dsde.rawls.RawlsException
+import org.broadinstitute.dsde.rawls.dataaccess._
+import org.broadinstitute.dsde.rawls.dataaccess.slick._
+import org.broadinstitute.dsde.rawls.model._
+import scala.concurrent.ExecutionContext
+import scala.util.{Try, Failure, Success}
+import scala.util.parsing.combinator.JavaTokenParsers
+
+case class SlickExpressionContext(workspaceContext: SlickWorkspaceContext, rootType: String, rootName: String)
+
+trait SlickExpressionParser extends JavaTokenParsers {
+  this: DriverComponent with EntityComponent with WorkspaceComponent with AttributeComponent =>
+  import driver.api._
+
+  // A fully built query is a root step that runs first the output of which is fed to
+  // a foldLeft over the steps the output of which is fed to the finalStep which will formulate
+  // the result. If rootStep is None, steps should also be empty and finalStep does all the work.
+  case class PipelineQuery(rootStep: Option[RootFunc], steps: List[PipeFunc], finalStep: FinalFunc)
```

Comment:
Would you mind reversing the order of these?  There are a lot of new concepts here, and I think it would be easier to understand them if introduced in dependency order.